### PR TITLE
Extract federation constants from BridgeConstants

### PR DESCRIPTION
**Pull request by julia-zack**
Extract federation constants from BridgeConstants to their own classes.
Create /federation package to have all federation-related classes.
Move new classes to /federation/constants package.

Fix now broken use of BridgeConstants

### CODE REVIEW

File: rskj-core/src/main/java/co/rsk/peg/constants/BridgeRegTestConstants.java [hunk 1]
```diff
@@ -34,25 +34,26 @@
...
-        BtcECKey.fromPrivate(HashUtil.keccak256("federator3".getBytes(StandardCharsets.UTF_8)))
-    ));
-    public static final List<BtcECKey> REGTEST_FEDERATION_PUBLIC_KEYS = Collections.unmodifiableList(REGTEST_FEDERATION_PRIVATE_KEYS.stream()
-        .map(key -> BtcECKey.fromPublicOnly(key.getPubKey()))
-        .collect(Collectors.toList()));
-
-    private static final BridgeRegTestConstants instance = new BridgeRegTestConstants(REGTEST_FEDERATION_PUBLIC_KEYS);
+    private static final BridgeRegTestConstants instance = new BridgeRegTestConstants();
+
+    public BridgeRegTestConstants() {
```
Comment (marcos-iov) [inline rskj-core/src/main/java/co/rsk/peg/constants/BridgeRegTestConstants.java:39]: If this class implements a Singleton pattern then this should be private

File: rskj-core/src/main/java/co/rsk/peg/constants/BridgeRegTestConstants.java [hunk 2]
```diff
@@ -34,25 +34,26 @@
...
+        BtcECKey federator1PrivateKey = BtcECKey.fromPrivate(HashUtil.keccak256("federator2".getBytes(StandardCharsets.UTF_8)));
+        BtcECKey federator2PrivateKey = BtcECKey.fromPrivate(HashUtil.keccak256("federator3".getBytes(StandardCharsets.UTF_8)));
+        List<BtcECKey> defaultFederationPrivateKeys = Collections.unmodifiableList(Arrays.asList(federator0PrivateKey, federator1PrivateKey, federator2PrivateKey));
+
+        return Collections.unmodifiableList(defaultFederationPrivateKeys.stream()
+                .map(key -> BtcECKey.fromPublicOnly(key.getPubKey()))
+                .collect(Collectors.toList()));
+    }
 
     public BridgeRegTestConstants(List<BtcECKey> federationPublicKeys) {
```
Comment (marcos-iov) [inline rskj-core/src/main/java/co/rsk/peg/constants/BridgeRegTestConstants.java:53]: Same here

File: rskj-core/src/main/java/co/rsk/peg/federation/constants/FederationRegTestConstants.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import org.bouncycastle.util.encoders.Hex;
+import org.ethereum.crypto.ECKey;
+import java.time.ZonedDateTime;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class FederationRegTestConstants extends FederationConstants {
+
+    public FederationRegTestConstants(List<BtcECKey> federationPublicKeys) {
```
Comment (marcos-iov) [inline rskj-core/src/main/java/co/rsk/peg/federation/constants/FederationRegTestConstants.java:14]: Should this one be a singleton as well?

File: rskj-core/src/main/java/co/rsk/peg/BridgeUtils.java [hunk 4]
```diff
@@ -460,7 +462,8 @@ public static Coin getCoinFromBigInteger(BigInteger value) throws BridgeIllegalA
     }
 
     private static boolean isFromFederationChangeAuthorizedSender(Transaction rskTx, BridgeConstants bridgeConfiguration, SignatureCache signatureCache) {
```
Comment (marcos-iov) [inline rskj-core/src/main/java/co/rsk/peg/BridgeUtils.java:464]: Should this receive FederationConstants directly maybe?

File: rskj-core/src/main/java/co/rsk/peg/federation/constants/FederationMainNetConstants.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        String erpFederator3StringPubKey = "02370a9838e4d15708ad14a104ee5606b36caaaaf739d833e67770ce9fd9b3ec80";
+        erpFedPubKeysList = Arrays.stream(new String[]
+            {erpFederator0StringPubKey, erpFederator1StringPubKey, erpFederator2StringPubKey, erpFederator3StringPubKey})
+            .map(hex -> BtcECKey.fromPublicOnly(Hex.decode(hex))).collect(Collectors.toList());
+        erpFedActivationDelay = 52_560; // 1 year in BTC blocks (considering 1 block every 10 minutes)
+
+        oldFederationAddress = "35JUi1FxabGdhygLhnNUEFG4AgvpNMgxK1";
+    }
+
+    public static FederationMainNetConstants getInstance() { return instance; }
```
Comment (apancorb) [inline rskj-core/src/main/java/co/rsk/peg/federation/constants/FederationMainNetConstants.java:69]: nit: consider placing the return statement on a new line for improved readability

File: rskj-core/src/main/java/co/rsk/peg/federation/constants/FederationRegTestConstants.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        erpFedActivationDelay = 500;
+
+        // Multisig address created in bitcoind with the following private keys:
+        // 47129ffed2c0273c75d21bb8ba020073bb9a1638df0e04853407461fdd9e8b83
+        // 9f72d27ba603cfab5a0201974a6783ca2476ec3d6b4e2625282c682e0e5f1c35
+        // e1b17fcd0ef1942465eee61b20561b16750191143d365e71de08b33dd84a9788
+        oldFederationAddress = "2N7ZgQyhFKm17RbaLqygYbS7KLrQfapyZzu";
+    }
+
+    public static FederationRegTestConstants getInstance(List<BtcECKey> federationPublicKeys) {
```
Comment (apancorb) [inline rskj-core/src/main/java/co/rsk/peg/federation/constants/FederationRegTestConstants.java:56]: Maybe there can be a `FederationConstantsFactory` class that exposes all these `getInstance` methods so they are more centralized in one place. Just a thought.

File: rskj-core/src/main/java/org/ethereum/config/SystemProperties.java [hunk 7]
```diff
@@ -186,9 +185,10 @@ public Constants getNetworkConstants() {
                     );
                     break;
                 case "regtest":
-                    constants = Constants.regtestWithFederation(
-                            getGenesisFederationPublicKeys().orElse(BridgeRegTestConstants.REGTEST_FEDERATION_PUBLIC_KEYS)
-                    );
+                    Optional<List<BtcECKey>> genesisFederationPublicKeysOptional = getGenesisFederationPublicKeys();
+                    constants = genesisFederationPublicKeysOptional
+                        .map(Constants::regtestWithFederation)
+                        .orElseGet(Constants::regtest);
```
Comment (apancorb) [inline rskj-core/src/main/java/org/ethereum/config/SystemProperties.java:191]:
```suggestion
                    constants = getGenesisFederationPublicKeys()
                        .map(Constants::regtestWithFederation)
                        .orElseGet(Constants::regtest);
```

File: rskj-core/src/main/java/co/rsk/peg/utils/BridgeEventLoggerImpl.java [hunk 8]
```diff
@@ -117,7 +118,9 @@ public void logCommitFederation(Block executionBlock, Federation oldFederation,
         String oldFederationBtcAddress = oldFederation.getAddress().toBase58();
         byte[] newFederationFlatPubKeys = flatKeysAsByteArray(newFederation.getBtcPublicKeys());
         String newFederationBtcAddress = newFederation.getAddress().toBase58();
-        long newFedActivationBlockNumber = executionBlock.getNumber() + this.bridgeConstants.getFederationActivationAge(activations);
+
+        FederationConstants federationConstants = this.bridgeConstants.getFederationConstants();
```
Comment (apancorb) [inline rskj-core/src/main/java/co/rsk/peg/utils/BridgeEventLoggerImpl.java:122]: nit: avoid using `this` to keep with the consistency of class

File: rskj-core/src/main/java/co/rsk/peg/federation/constants/FederationRegTestConstants.java (hunk 6 above)
Comment (marcos-iov) [reply rskj-core/src/main/java/co/rsk/peg/federation/constants/FederationRegTestConstants.java:56]: This is a bit strange. A singleton pattern that receives params. Basically those keys will only be used if the instance is not created, if there is already an instance then they will be ignored. That can be misleading

File: rskj-core/src/main/java/org/ethereum/config/SystemProperties.java (hunk 7 above)
Comment (j-zack) [reply rskj-core/src/main/java/org/ethereum/config/SystemProperties.java:191]: good one, thanks!

Comment (julia-zack) [reply rskj-core/src/main/java/org/ethereum/config/SystemProperties.java:191]: good one, thanks!

File: rskj-core/src/main/java/co/rsk/peg/federation/constants/FederationRegTestConstants.java (hunk 6 above)
Comment (julia-zack) [reply rskj-core/src/main/java/co/rsk/peg/federation/constants/FederationRegTestConstants.java]: i have undone the singleton pattern implementation, thanks guys

**Review by marcos-iov — APPROVED** [review]

**Pull request merged**